api/include/opentelemetry/nostd/list.h

```diff
@@ -0,0 +1,250 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <iterator>
+#include <memory>
+
+#include "opentelemetry/nostd/utility.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace nostd
+{
+/* Non-standard implementation of the list data structure. */
+template <class T>
+class list
+{
+public:
+  typedef T value_type;
+
+  /* Custom iterator class*/
+  class iterator
+  {
+  public:
+    typedef std::forward_iterator_tag iterator_category;
+    iterator(T *ptr) : ptr_(ptr) {}
+    iterator operator++()
+    {
+      ptr_++;
+      return ptr_;
+    }
+    T &operator*() { return *ptr_; }
+    T *operator->() { return ptr_; }
+    bool operator==(const iterator &other) { return ptr_ == other.ptr_; }
+    bool operator!=(const iterator &other) { return ptr_ != other.ptr_; }
+
+  private:
+    T *ptr_;
+  };
+
+  list()
+  {
+    size_ = 0;
+    data_ = NULL;
+  };
+
+  /* Constructs class with passed in element. */
+  list(T element)
+  {
+    size_ = 1;
+
+    T *data_ = new T[size_];
+
+    std::memcpy(data_, &element, sizeof(T));
+  };
+
+  /* Constructs class with the passed in initializer_list. */
+  list(std::initializer_list<T> init_list)
+  {
+    size_ = init_list.size();
+
+    T *temp = new T[size_];
+
+    std::memcpy(temp, std::begin(init_list), init_list.size() * sizeof(T));
+
+    data_ = temp;
+  }
+
+  /* Returns the element at the front of the list. */
+  T front()
+  {
+    if (size_ == 0)
+    {
+      return T();
+    }
```

Comment:
Just FYI, not everything is going to be default constructable. Consider just returning data_[0]. Requiring the caller to make sure the list isn't empty is reasonable, and since the list could contain T() they can't e.g. use it as a sentinel value and stop when they see it.

Reply:
Changed from T() to data_[0].